### ISSUE TITLE
Consul Connect `upstream connect error or disconnect/reset before headers`

### ISSUE THREAD

**Issue by Crizstian** [hashicorp/nomad#8002]
### Nomad version
Output from `nomad version`
```bash
root@sfo-consul-server:/vagrant/provision/terraform/tf_cluster/primary# nomad version
Nomad v0.11.2 (807cfebe90d56f9e5beec3e72936ebe86acc8ce3)

Consul v1.7.3
Protocol 2 spoken by default, understands 2 to 3 (agent will automatically use protocol >2 when speaking to compatible agents)
```
### Operating system and Environment details
Linux
### Issue

```
Nomad task doesn't connect using the sidecar upstreams; it gives the following error.

upstream connect error or disconnect/reset before headers. reset reason: connection termination/tmp #
```

### Test

Environment

- Nomad with TLS
- Consul with TLS and ACL enabled

once services are deployed
- get inside the container

- made simple curl to upstream defined

```bash
root@sfo-consul-server:/vagrant/provision/terraform/tf_cluster/primary# nomad alloc exec -task booking-api c6207633 sh

/tmp # curl $PAYMENT_URL/ping
upstream connect error or disconnect/reset before headers. reset reason: connection termination
```

```bash
/tmp # curl -I -v $PAYMENT_URL/ping
*   Trying 127.0.0.1:8080...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8080 (#0)
> HEAD /ping HTTP/1.1
> Host: 127.0.0.1:8080
> User-Agent: curl/7.67.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 503 Service Unavailable
HTTP/1.1 503 Service Unavailable
< content-length: 95
content-length: 95
< content-type: text/plain
content-type: text/plain
< date: Mon, 18 May 2020 19:07:15 GMT
date: Mon, 18 May 2020 19:07:15 GMT
< server: envoy
server: envoy

<
* Connection #0 to host 127.0.0.1 left intact
```

sidecar-proxy logs

```bash
root@sfo-consul-server:/vagrant/provision/terraform/tf_cluster/primary# nomad logs -stderr c6207633 connect-proxy-booking-api
[2020-05-18 18:09:32.183][1][info][main] [source/server/server.cc:238] initializing epoch 0 (hot restart version=disabled)
[2020-05-18 18:09:32.184][1][info][main] [source/server/server.cc:240] statically linked extensions:
[2020-05-18 18:09:32.184][1][info][main] [source/server/server.cc:242]   access_loggers: envoy.file_access_log,envoy.http_grpc_access_log
[2020-05-18 18:09:32.184][1][info][main] [source/server/server.cc:245]   filters.http: envoy.buffer,envoy.cors,envoy.csrf,envoy.ext_authz,envoy.fault,envoy.filters.http.dynamic_forward_proxy,envoy.filters.http.grpc_http1_reverse_bridge,envoy.filters.http.header_to_metadata,envoy.filters.http.jwt_authn,envoy.filters.http.original_src,envoy.filters.http.rbac,envoy.filters.http.tap,envoy.grpc_http1_bridge,envoy.grpc_json_transcoder,envoy.grpc_web,envoy.gzip,envoy.health_check,envoy.http_dynamo_filter,envoy.ip_tagging,envoy.lua,envoy.rate_limit,envoy.router,envoy.squash
[2020-05-18 18:09:32.184][1][info][main] [source/server/server.cc:248]   filters.listener: envoy.listener.original_dst,envoy.listener.original_src,envoy.listener.proxy_protocol,envoy.listener.tls_inspector
[2020-05-18 18:09:32.184][1][info][main] [source/server/server.cc:251]   filters.network: envoy.client_ssl_auth,envoy.echo,envoy.ext_authz,envoy.filters.network.dubbo_proxy,envoy.filters.network.mysql_proxy,envoy.filters.network.rbac,envoy.filters.network.sni_cluster,envoy.filters.network.thrift_proxy,envoy.filters.network.zookeeper_proxy,envoy.http_connection_manager,envoy.mongo_proxy,envoy.ratelimit,envoy.redis_proxy,envoy.tcp_proxy
[2020-05-18 18:09:32.184][1][info][main] [source/server/server.cc:253]   stat_sinks: envoy.dog_statsd,envoy.metrics_service,envoy.stat_sinks.hystrix,envoy.statsd
[2020-05-18 18:09:32.184][1][info][main] [source/server/server.cc:255]   tracers: envoy.dynamic.ot,envoy.lightstep,envoy.tracers.datadog,envoy.tracers.opencensus,envoy.zipkin
[2020-05-18 18:09:32.184][1][info][main] [source/server/server.cc:258]   transport_sockets.downstream: envoy.transport_sockets.alts,envoy.transport_sockets.tap,raw_buffer,tls
[2020-05-18 18:09:32.184][1][info][main] [source/server/server.cc:261]   transport_sockets.upstream: envoy.transport_sockets.alts,envoy.transport_sockets.tap,raw_buffer,tls
[2020-05-18 18:09:32.184][1][info][main] [source/server/server.cc:267] buffer implementation: old (libevent)
[2020-05-18 18:09:32.185][1][warning][misc] [source/common/protobuf/utility.cc:199] Using deprecated option 'envoy.api.v2.Cluster.hosts' from file cds.proto. This configuration will be removed from Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated for details.
[2020-05-18 18:09:32.188][1][info][main] [source/server/server.cc:322] admin address: 127.0.0.1:19001
[2020-05-18 18:09:32.188][1][info][main] [source/server/server.cc:432] runtime: layers:
  - name: static_layer
    static_layer:
      envoy.deprecated_features:envoy.api.v2.Cluster.tls_context: true
      envoy.deprecated_features:envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager.Tracing.operation_name: true
      envoy.deprecated_features:envoy.config.trace.v2.ZipkinConfig.HTTP_JSON_V1: true
[2020-05-18 18:09:32.188][1][warning][runtime] [source/common/runtime/runtime_impl.cc:497] Skipping unsupported runtime layer: name: "static_layer"
static_layer {
  fields {
    key: "envoy.deprecated_features:envoy.api.v2.Cluster.tls_context"
    value {
      bool_value: true
    }
  }
  fields {
    key: "envoy.deprecated_features:envoy.config.filter.network.http_connection_manager.v2.HttpConnectionManager.Tracing.operation_name"
    value {
      bool_value: true
    }
  }
  fields {
    key: "envoy.deprecated_features:envoy.config.trace.v2.ZipkinConfig.HTTP_JSON_V1"
    value {
      bool_value: true
    }
  }
}

[2020-05-18 18:09:32.188][1][info][config] [source/server/configuration_impl.cc:61] loading 0 static secret(s)
[2020-05-18 18:09:32.188][1][info][config] [source/server/configuration_impl.cc:67] loading 1 cluster(s)
[2020-05-18 18:09:32.196][1][info][upstream] [source/common/upstream/cluster_manager_impl.cc:144] cm init: initializing cds
[2020-05-18 18:09:32.199][1][info][config] [source/server/configuration_impl.cc:71] loading 0 listener(s)
[2020-05-18 18:09:32.199][1][info][config] [source/server/configuration_impl.cc:96] loading tracing configuration
[2020-05-18 18:09:32.199][1][info][config] [source/server/configuration_impl.cc:116] loading stats sink configuration
[2020-05-18 18:09:32.199][1][info][main] [source/server/server.cc:516] starting main dispatch loop
[2020-05-18 18:09:32.207][1][info][upstream] [source/common/upstream/cluster_manager_impl.cc:489] add/update cluster local_app during init
[2020-05-18 18:09:32.212][1][info][upstream] [source/common/upstream/cluster_manager_impl.cc:489] add/update cluster payment-api.default.sfo.internal.be0da89e-6ad5-3069-8e94-8db658f51b5b.consul during init
[2020-05-18 18:09:32.218][1][info][upstream] [source/common/upstream/cluster_manager_impl.cc:489] add/update cluster notification-api.default.sfo.internal.be0da89e-6ad5-3069-8e94-8db658f51b5b.consul during init
[2020-05-18 18:09:32.218][1][info][upstream] [source/common/upstream/cluster_manager_impl.cc:124] cm init: initializing secondary clusters
[2020-05-18 18:09:32.219][1][info][upstream] [source/common/upstream/cluster_manager_impl.cc:148] cm init: all clusters initialized
[2020-05-18 18:09:32.219][1][info][main] [source/server/server.cc:500] all clusters initialized. initializing init manager
[2020-05-18 18:09:32.223][1][info][upstream] [source/server/lds_api.cc:60] lds: add/update listener 'public_listener:0.0.0.0:30091'
[2020-05-18 18:09:32.225][1][info][upstream] [source/server/lds_api.cc:60] lds: add/update listener 'payment-api:127.0.0.1:8080'
[2020-05-18 18:09:32.226][1][info][upstream] [source/server/lds_api.cc:60] lds: add/update listener 'notification-api:127.0.0.1:8081'
[2020-05-18 18:09:32.227][1][info][upstream] [source/server/lds_api.cc:60] lds: add/update listener 'exposed_path_ping:0.0.0.0:25804'
[2020-05-18 18:09:32.227][1][info][config] [source/server/listener_manager_impl.cc:761] all dependencies initialized. starting workers
[2020-05-18 18:24:32.222][1][info][main] [source/server/drain_manager_impl.cc:63] shutting down parent after drain
```

How can I debug sidecar / upstream connections ?

Consul Logs doesn't say nothing
Nomad Logs doesn't say nothing

Envoy issue ???

### Job file (if appropriate)

cinemas hcl file
```hcl
job "cinemas1" {

  group "payment-api" {
    count = 1

    network {
      mode = "bridge"
      
      port "healthcheck" {
        to = -1
      }
    }

    service {
      name = "payment-api"
      port = "3000"

      check {
        name     = "payment-api-health"
        port     = "healthcheck"
        type     = "http"
        protocol = "http"
        path     = "/ping"
        interval = "5s"
        timeout  = "2s"
        expose   = true
      }

      connect {
        sidecar_service {}
      }
    }

    task "payment-api" {
      driver = "docker"

      config {
        image = "crizstian/payment-service-go:v0.4"
      }

      env {
        DB_SERVERS      = "mongodb1.service.consul:27017,mongodb2.service.consul:27018,mongodb3.service.consul:27019"
        SERVICE_PORT    = "3000"
        CONSUL_IP       = "consul.service.consul"
        CONSUL_SCHEME   = "https"
        CONSUL_HTTP_SSL = "true"
      }

      resources {
        cpu    = 50
        memory = 50
      }
    }
  }

  group "notification-api" {
    count = 1
    
    network {
      mode = "bridge"

      port "healthcheck" {
        to = -1
      }
    }

      service {
      name = "notification-api"
      port = "3001"

      check {
        name     = "notification-api-health"
        port     = "healthcheck"
        type     = "http"
        protocol = "http"
        path     = "/ping"
        interval = "5s"
        timeout  = "2s"
        expose   = true
      }

      connect {
        sidecar_service {}
      }
    }

    task "notification-api" {
      driver = "docker"

      config {
        image   = "crizstian/notification-service-go:v0.4"
      }

      env {
        SERVICE_PORT    = "3001"
        CONSUL_IP       = "consul.service.consul"
        CONSUL_SCHEME   = "https"
        CONSUL_HTTP_SSL = "true"
      }

      resources {
        cpu    = 50
        memory = 50
      }
    }
  }


  group "booking-api" {
    count = 1

    network {
      mode = "bridge"

      port "http" {
        static = 3002
        to     = 3002
      }

      port "healthcheck" {
        to = -1
      }
    }

    service {
      name = "booking-api"
      port = "http"
      tags = ["cinemas-project"]

      check {
        name     = "booking-api-health"
        port     = "healthcheck"
        type     = "http"
        protocol = "http"
        path     = "/ping"
        interval = "10s"
        timeout  = "3s"
        expose   = true
      }

      connect {
        sidecar_service {
          proxy {
            upstreams {
               destination_name = "payment-api"
               local_bind_port = 8080
            }
            upstreams {
               destination_name = "notification-api"
               local_bind_port = 8081
            }
          }
        }
      }
    }

    task "booking-api" {
      driver = "docker"

      config {
        image   = "crizstian/booking-service-go:v0.4"
      }

      env {
        SERVICE_PORT     = "3002"
        DB_SERVERS       = "mongodb1.query.consul:27017,mongodb2.query.consul:27018,mongodb3.query.consul:27019"

        CONSUL_IP        = "consul.service.consul"
        CONSUL_SCHEME    = "https"
        CONSUL_HTTP_SSL  = "true"

        PAYMENT_URL      = "http://${NOMAD_UPSTREAM_ADDR_payment_api}"
        NOTIFICATION_URL = "http://${NOMAD_UPSTREAM_ADDR_notification_api}"
      }

      resources {
        cpu    = 50
        memory = 50
      }
    }
  }
}
```

### Nomad Client logs (if appropriate)
If possible please post relevant logs in the issue.
```bash
root@sfo-consul-server:/vagrant/provision/terraform/tf_cluster/primary# nomad status cinemas1
ID            = cinemas1
Name          = cinemas1
Submit Date   = 2020-05-18T18:09:30Z
Type          = service
Priority      = 50
Datacenters   = sfo-ncv
Namespace     = default
Status        = running
Periodic      = false
Parameterized = false

Summary
Task Group        Queued  Starting  Running  Failed  Complete  Lost
booking-api       0       0         1        0       0         0
notification-api  0       0         1        0       0         0
payment-api       0       0         1        0       1         0

Latest Deployment
ID          = 04d7f48e
Status      = successful
Description = Deployment completed successfully

Deployed
Task Group        Desired  Placed  Healthy  Unhealthy  Progress Deadline
booking-api       1        1       1        0          2020-05-18T18:19:51Z
notification-api  1        1       1        0          2020-05-18T18:19:45Z
payment-api       1        1       1        0          2020-05-18T18:19:44Z

Allocations
ID        Node ID   Task Group        Version  Desired  Status    Created     Modified
c6207633  c0e5ca0a  booking-api       3        run      running   21m32s ago  21m12s ago
a18b48ee  c0e5ca0a  notification-api  3        run      running   26m13s ago  21m17s ago
0639aab9  c0e5ca0a  payment-api       3        run      running   32m22s ago  21m18s ago
ff3d88c0  c0e5ca0a  payment-api       0        stop     complete  33m22s ago  32m17s ago
```

### FUN FACT

countdash.hcl
```
job "countdash" {

  group "api" {
    count = 1

    network {
      mode = "bridge"

      port "healthcheck" {
        to = -1
      }
    }

    service {
      name = "count-api"
      port = "9001"

      check {
        name     = "api-health"
        port     = "healthcheck"
        type     = "http"
        protocol = "http"
        path     = "/health"
        interval = "10s"
        timeout  = "3s"
        expose   = true
      }

      connect {
        sidecar_service {}
      }
    }

    task "web" {
      driver = "docker"

      config {
        image = "hashicorpnomad/counter-api:v1"
      }

      resources {
        cpu    = 50
        memory = 50
      }
    }
  }

  group "dashboard" {
    count = 1
    
    network {
      mode = "bridge"

      port "http" {
        static = 9002
        to     = 9002
      }

      port "healthcheck" {
        to = -1
      }
    }

    service {
      name = "count-dashboard"
      port = "http"

      check {
        name     = "dashboard-health"
        port     = "healthcheck"
        type     = "http"
        protocol = "http"
        path     = "/health"
        interval = "10s"
        timeout  = "3s"
        expose   = true
      }

      connect {
        sidecar_service {
          proxy {
            upstreams {
              destination_name = "count-api"
              local_bind_port  = 8080
            }
          }
        }
      }
    }

    task "dashboard" {
      driver = "docker"

      env {
        COUNTING_SERVICE_URL = "http://${NOMAD_UPSTREAM_ADDR_count_api}"
      }

      config {
        image = "hashicorpnomad/counter-dashboard:v1"
      }

      resources {
        cpu    = 50
        memory = 50
      }
    }
  }
}
```

Countdash works, with intentions enabled with intension denied also worked and without any intentions.

why count dash works and cinemas doesn't if they both have same configuration pattern

### Extras

Consul Configs applied

```bash
root@sfo-consul-server:/vagrant/provision/terraform/tf_cluster/primary# consul config list -kind service-defaults
booking-service
count-api
count-dashboard
notification-api
payment-api
```

### Upstream services working and marked as healthy in Consul
![Captura de Pantalla 2020-05-18 a la(s) 13 39 32](https://user-images.githubusercontent.com/11022466/82248033-1853ef80-990d-11ea-8081-64b51f1037cb.png)

![Captura de Pantalla 2020-05-18 a la(s) 13 40 30](https://user-images.githubusercontent.com/11022466/82248074-2ace2900-990d-11ea-870b-a2c1d58e786d.png)

Booking sidecar proxy working 

![Captura de Pantalla 2020-05-18 a la(s) 13 41 11](https://user-images.githubusercontent.com/11022466/82248156-46d1ca80-990d-11ea-84dd-78707151dfeb.png)

Logs and other artifacts may also be sent to: nomad-oss-debug@hashicorp.com

Please link to your Github issue in the email and reference it in the subject
line:

> To: nomad-oss-debug@hashicorp.com
>
> Subject: GH-1234: Errors garbage collecting allocs

Emails sent to that address are readable by all HashiCorp employees but are *not* publicly visible.

**Comment by shoenig** [comment]
Hey @Crizstian just taking a quick first glance, I'm wondering if using Consul DNS addresses here is causing problems. Remember, with bridge networking the tasks are running in a network namespace isolated from the host. You could try using the `template` stanza to resolve those addresses beforehand, or configure a task that can resolve DNS requests, etc. 

(FWIW the upcoming Nomad v0.12 release will include some DNS features that should make this a little easier)

Is there any useful information in the output from `docker logs` for each of these tasks?

**Comment by Crizstian** [comment]
I could make it work now, and I found where was the issue.

I believe is a consul and consul terraform provider issue.

Problem is that I was creating my consul intentions with terraform, so this intentions was created first, then I made the nomad deployments. all of this looks ok to me, but, in reality the consul intentions were not doing its job, they were created and they showed in the consul ui, but this were not allowing the sidecars to communicate.

i tear down all my configs and start fresh again first without tls since I thought it was still a tls issue, and then deployed my stuff and finally create the intentions. And it works.

i setup again the environment with consul tls enabled and acls enabled, deployed the nomad jobs, and then I created the intentions manually again and ta da, it works.

then I did a couple of more tests, I block the connections in the intentions, and the communications still was happening, I allowed the intentions again and communications works as expected I blocked all service to service and still the communications was happening.

in conclusion I believe there is a bug in consul intentions, sidecars and the consul terraform provider (probably)

i will open a new issue in the consul repo, and close this one since from nomad side everything is working.
